src/compressed_tensors/compressors/marlin_24.py

```diff
@@ -0,0 +1,267 @@
+        :param weight: tensor to check for sparsity structure
+        :param num_rows_to_sample: number of rows to check the sparsity structure of
+        :return: True if all sampled rows match the 2:4 sparsity structure, raises
+            ValueError otherwise
+        """
+        BLOCK_SIZE = 4
+        MAX_NON_ZEROS = 2
+
+        weight = weight.contiguous()
+
+        num_rows, num_cols = weight.shape
+        sampled_row_idxs = random.choices(range(num_rows), k=num_rows_to_sample)
+
+        non_24_segments = 0
+        for i in sampled_row_idxs:
+            for j in range(0, num_cols - BLOCK_SIZE, BLOCK_SIZE):
+                block = weight[i, j : j + BLOCK_SIZE]
+                num_nonzero = torch.count_nonzero(block)
+                if num_nonzero > MAX_NON_ZEROS:
+                    non_24_segments += 1
+
+        if non_24_segments > 0:
+            raise ValueError(
+                "Marlin24 Compressor is only compatible with weights that have "
+                f"a 2:4 sparsity structure. Found {non_24_segments} segments in {name} "
+                "that do not match the expected structure."
+            )
+
+        return True
+
+    def compress(
+        self,
+        model_state: Dict[str, Tensor],
+        model_quant_args: Dict[str, QuantizationArgs],
+        **kwargs,
+    ) -> Dict[str, Tensor]:
+        """
+        Compresses a quantized state_dict with 2:4 sparsity structure for inference
+        with the Marlin24 kernel
+
+        :param model_state: state dict of uncompressed model
+        :param model_quant_args: quantization args for each quantized weight, needed for
+        quantize function to calculate bit depth
+        :return: compressed state dict
+        """
+        self.validate_quant_compatability(model_quant_args)
+
+        compressed_dict = {}
+        weight_suffix = ".weight"
+        _LOGGER.debug(
+            f"Compressing model with {len(model_state)} parameterized layers..."
+        )
+
+        for name, value in tqdm(model_state.items(), desc="Compressing model"):
+            if name.endswith(weight_suffix):
+                prefix = name[: -(len(weight_suffix))]
+                scale = model_state.get(merge_names(prefix, "weight_scale"), None)
+                zp = model_state.get(merge_names(prefix, "weight_zero_point"), None)
+                if scale is not None:  # weight is quantized, compress it
+
+                    # Marlin24 kernel requires float16 inputs
+                    scale = scale.to(torch.float16)
+                    value = value.to(torch.float16)
+
+                    # quantize weight, keeping it as a float16 for now
+                    quant_args = model_quant_args[prefix]
+                    value = quantize(
+                        x=value, scale=scale, zero_point=zp, args=quant_args
+                    )
+
+                    # compress based on sparsity structure
+                    self.validate_sparsity_structure(prefix, value)
```

Comment:
Ok i see why you have two validators bc of here

Reply:
Yeah one of the validators is for the quantization args, the other checks each weight individually